node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.SignedHeader](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID)
+	if err = bExService.sub.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting subscriber: %w", err)
+	}
+	if _, err := bExService.sub.Subscribe(); err != nil {
+		return fmt.Errorf("error while subscribing: %w", err)
+	}
+
+	if err = bExService.blockStore.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting header store: %w", err)
+	}
+
+	_, _, network := bExService.p2p.Info()
+	if bExService.p2pServer, err = newP2PServer(bExService.p2p.Host(), bExService.blockStore, network); err != nil {
+		return err
+	}
+	if err = bExService.p2pServer.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting p2p server: %w", err)
+	}
+
+	peerIDs := bExService.p2p.PeerIDs()
+	if bExService.ex, err = newP2PExchange(bExService.p2p.Host(), peerIDs, network, bExService.genesis.ChainID, bExService.p2p.ConnectionGater()); err != nil {
+		return err
+	}
+	if err = bExService.ex.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting exchange: %w", err)
+	}
+
+	if bExService.syncer, err = newSyncer(bExService.ex, bExService.blockStore, bExService.sub, goheadersync.WithBlockTime(bExService.conf.BlockTime)); err != nil {
+		return err
+	}
+
+	// Check if the headerstore is not initialized and try initializing
+	if bExService.blockStore.Height() > 0 {
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		bExService.syncerStatus.started = true
+		return nil
+	}
+
+	// Look to see if trusted hash is passed, if not get the genesis header
+	var trustedHeader *types.SignedHeader
+	// Try fetching the trusted header from peers if exists
+	if len(peerIDs) > 0 {
+		if bExService.conf.TrustedHash != "" {
+			trustedHashBytes, err := hex.DecodeString(bExService.conf.TrustedHash)
+			if err != nil {
+				return fmt.Errorf("failed to parse the trusted hash for initializing the headerstore: %w", err)
+			}
+
+			if trustedHeader, err = bExService.ex.Get(bExService.ctx, header.Hash(trustedHashBytes)); err != nil {
+				return fmt.Errorf("failed to fetch the trusted header for initializing the headerstore: %w", err)
+			}
+		} else {
+			// Try fetching the genesis header if available, otherwise fallback to signed headers
+			if trustedHeader, err = bExService.ex.GetByHeight(bExService.ctx, uint64(bExService.genesis.InitialHeight)); err != nil {
+				// Full/light nodes have to wait for aggregator to publish the genesis header
+				// proposing aggregator can init the store and start the syncer when the first header is published
+				bExService.logger.Info("failed to fetch the genesis header", "error", err)
+			}
+		}
+	}
+	go bExService.tryInitBlockStoreAndStartSyncer(bExService.ctx, trustedBlock)
```

Comment:
what is the rationale behind not blocking on this?

Reply:
Talked to Ganesh about it and made this blocking.